File: 2/calc-2.js
const fs = require("fs");
const { reduce } = require("lodash");

const input = fs.readFileSync("./input-1.txt").toString();

const scores = {
  A: 1, // rock
  B: 2, // paper
  C: 3, // scissors
  X: 0, // lose
  Y: 3, // draw
  Z: 6, // win
};

const rounds = input.split("\n").slice(0, -1);

const points = rounds.map((round) => {
  let points = 0;
  const [team1, team2] = round.split(" ");
  points += scores[team2];
  if (team2 === "Y") {
    // draw
    points += scores[team1];
  }
  if (team2 === "Z") {
    // win
    const morePoints = (scores[team1] + 1) % 3;
    points += morePoints === 0 ? 1 : morePoints;
  }
  if (team2 === "X") {
    // lose
    const morePoints = (scores[team1] - 1 + 3) % 3;
    points += morePoints === 0 ? 3 : morePoints;
  }
  return points;
});

const totalPoints = points.reduce((acc, curr) => acc + curr, 0);

console.log(totalPoints);

File: 2/calc-2.clj
(defn parse [xs]
  (map #(Integer/parseInt %) xs))

(defn calc-score [scores opponent us]
  (let [points-from-result (scores us)
        points-from-play (case us
                           "X" ;; lose
                           (scores (case opponent
                                     "A" "C"
                                     "B" "A"
                                     "C" "B"))
                           "Y" ;; draw
                           (scores opponent)
                           "Z" ;; win
                           (scores (case opponent
                                     "A" "B"
                                     "B" "C"
                                     "C" "A")))]
    (+ points-from-play points-from-result)))

(let [input (slurp "input-1.txt")
      scores {"A" 1 "B" 2 "C" 3 "X" 0 "Y" 3 "Z" 6}]
  (as-> input v
    (str/split v #"\n") ;; split into lines
    (map #(str/split % #" ") v)
    (map #(calc-score scores (first %) (second %)) v)
    (reduce + v)
    (println v)))
